Sync/BatchingClient.swift

```diff
@@ -0,0 +1,263 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Alamofire
+import Shared
+import XCGLogger
+import Deferred
+
+public class BatchingNotSupported: MaybeErrorType {
+    public let response: POSTResult
+
+    public var description: String {
+        return "Sync server does not support batching."
+    }
+
+    public init(response: POSTResult) {
+        self.response = response
+    }
+}
+
+private let log = Logger.syncLogger
+
+typealias BatchUploadFunction = (lines: [String], ifUnmodifiedSince: Timestamp?, queryParams: [NSURLQueryItem]?) -> Deferred<Maybe<StorageResponse<POSTResult>>>
+
+public class Sync15BatchClient<T: CleartextPayloadJSON> {
+    private let config: InfoConfiguration
+
+    private var records: [Record<T>] = []
+    private let ifUnmodifiedSince: Timestamp?
+    private let uploader: BatchUploadFunction
+    private let serializeRecord: (Record<T>) -> String?
+
+    // TODO: use I-U-S.
+    // Each time we do the storage operation, we might receive a backoff notification.
+    // For a success response, this will be on the subsequent request, which means we don't
+    // have to worry about handling successes and failures mixed with backoffs here.
+
+    init(config: InfoConfiguration, ifUnmodifiedSince: Timestamp? = nil, serializeRecord: (Record<T>) -> String?, uploader: BatchUploadFunction) {
+        self.config = config
+        self.ifUnmodifiedSince = ifUnmodifiedSince
+        self.uploader = uploader
+        self.serializeRecord = serializeRecord
+    }
+
+    public func addRecords(records: [Record<T>]) {
+        log.debug("Adding \(records.count) records into batch")
+        self.records += records
+    }
+
+    public func commit(onCollectionUploaded: (POSTResult -> Void)) -> Success {
+
+        if records.isEmpty {
+            return succeed()
+        }
+
+        // Need to deduce how big the data is we're sending across using this reducer
+        let sizeReducer: (total: ByteCount, record: Record<T>) -> ByteCount = { total, record in
+            return total + (self.serializeRecord(record) ?? "").utf8.count
```

Comment:
This is really painful. You're serializing every record all the way through to bytes… twice.